horovod/torch/mpi_ops_v2.cc

```diff
@@ -0,0 +1,324 @@
+  auto hvd_tensor = std::make_shared<TorchTensor>(tensor);
+  auto hvd_context = std::make_shared<TorchOpContext>(device, output);
+  std::shared_ptr<Tensor> hvd_output = nullptr;
+  if (horovod_rank() == root_rank) {
+    if (tensor.data_ptr() != output.data_ptr()) {
+      output.copy_(tensor);
+    }
+  } else {
+    hvd_output = std::make_shared<TorchTensor>(output);
+  }
+
+  auto handle = handle_manager.AllocateHandle();
+  auto enqueue_result =
+      EnqueueTensorBroadcast(hvd_context, hvd_tensor, hvd_output, root_rank,
+                             ready_event, GetOpName("broadcast", name, handle),
+                             device, [handle](const Status& status) {
+                               handle_manager.MarkDone(handle, status);
+                             });
+  ThrowIfError(enqueue_result);
+
+  return handle;
+}
+
+#if HAVE_CUDA
+int DoBroadcastCudaOnCPU(at::Tensor tensor, at::Tensor output, int root_rank,
+                         const std::string& name) {
+  ThrowIfError(common::CheckInitialized());
+
+  // Make async copy of input tensor to CPU tensor and record completion event.
+  auto device = GetDeviceID(tensor);
+  auto cpu_buffer =
+      tensor.to(at::Device(at::DeviceType::CPU), /*non_blocking=*/true);
+  auto hvd_cpu_buffer = std::make_shared<TorchTensor>(cpu_buffer);
+  auto ready_event = RecordReadyEvent(device);
+
+  auto hvd_context =
+      std::make_shared<TorchOpContext>(CPU_DEVICE_ID, cpu_buffer);
+
+  auto handle = handle_manager.AllocateHandle();
+  auto enqueue_result = EnqueueTensorBroadcast(
+      hvd_context, hvd_cpu_buffer, hvd_cpu_buffer, root_rank, ready_event,
+      GetOpName("broadcast", name, handle), CPU_DEVICE_ID,
+      [handle, cpu_buffer, output](const Status& status) mutable {
+        output.copy_(cpu_buffer);
+        handle_manager.MarkDone(handle, status);
+      });
+  ThrowIfError(enqueue_result);
+
+  return handle;
+}
+#endif
+
+int PollHandle(int handle) { return handle_manager.PollHandle(handle) ? 1 : 0; }
+
+void WaitAndClear(int handle) {
+  while (!handle_manager.PollHandle(handle)) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(1));
+  }
+  auto status = handle_manager.ReleaseHandle(handle);
+  ThrowIfError(*status);
+}
+
+PYBIND11_MODULE(mpi_lib_v2, m) {
+  // allreduce
+  m.def("horovod_torch_allreduce_async_torch_IntTensor", &DoAllreduce);
+  m.def("horovod_torch_allreduce_async_torch_LongTensor", &DoAllreduce);
+  m.def("horovod_torch_allreduce_async_torch_FloatTensor", &DoAllreduce);
+  m.def("horovod_torch_allreduce_async_torch_DoubleTensor", &DoAllreduce);
+#if HOROVOD_GPU_ALLREDUCE
+  m.def("horovod_torch_allreduce_async_torch_cuda_IntTensor", &DoAllreduce);
+  m.def("horovod_torch_allreduce_async_torch_cuda_LongTensor", &DoAllreduce);
+  m.def("horovod_torch_allreduce_async_torch_cuda_FloatTensor", &DoAllreduce);
+  m.def("horovod_torch_allreduce_async_torch_cuda_DoubleTensor", &DoAllreduce);
+#elif !HOROVOD_GPU_ALLREDUCE && HAVE_CUDA
```

Comment:
Nit: Isn't checking `!HOROVOD_GPU_ALLREDUCE` here redundant given that `HOROVOD_GPU_ALLREDUCE` evaluated to false above?

Same for allgather and broadcast.

Reply:
Good point :-)